src/Microsoft.NetCore.Analyzers/Core/Security/DoNotReferSelfInSerializableClass.cs

```diff
@@ -148,6 +106,76 @@ public sealed override void Initialize(AnalysisContext context)
                             }
                         });
 
+                    /// <summary>
+                    /// Traverse from point to its descendants, save the information into a directed graph.
+                    /// </summary>
+                    /// <param name="point">The initial point</param>
+                    void DrawGraph(ISymbol point)
+                    {
+                        // If the point has been visited, return.
+                        if (forwardGraph.Keys.Contains(point))
```

Comment:
If two threads both check `forwardGraph.Keys.Contains(point)` for the same `point` at the same time and both get back false, is the rest of the code thread safe?

One way to make both checking and marking-as-visited atomic is to take advantage of ConcurrentDictionary's `TryAdd()`'s return value when adding to `forwardGraph`.